src/main/java/io/cdap/plugin/sap/transformer/ODataEntryToRecordTransformer.java

```diff
@@ -131,13 +146,208 @@ private Object extractValue(String fieldName, Object value, Schema schema) {
+          break;
+        case LINESTRING:
+          Schema.Field lineStringsField = schema.getField(SapODataConstants.GEO_COLLECTION_LINE_STRINGS_FIELD_NAME);
+          Schema lineStringSchema = lineStringsField.getSchema().getComponentSchema();
+          lineStrings.add(extractGeospatial(fieldName, g, lineStringSchema));
+          break;
+        case POLYGON:
+          Schema.Field polygonsField = schema.getField(SapODataConstants.GEO_COLLECTION_POLYGONS_FIELD_NAME);
+          Schema polygonSchema = polygonsField.getSchema().getComponentSchema();
+          polygons.add(extractGeospatial(fieldName, g, polygonSchema));
+          break;
+        case MULTIPOINT:
+          Schema.Field multiPointsField = schema.getField(SapODataConstants.GEO_COLLECTION_MULTI_POINTS_FIELD_NAME);
+          Schema multiPointSchema = multiPointsField.getSchema().getComponentSchema();
+          multiPoints.add(extractGeospatial(fieldName, g, multiPointSchema));
+          break;
+        case MULTILINESTRING:
+          Schema.Field multiLsField = schema.getField(SapODataConstants.GEO_COLLECTION_MULTI_LINE_STRINGS_FIELD_NAME);
+          Schema multiLineStringSchema = multiLsField.getSchema().getComponentSchema();
+          multiLineStrings.add(extractGeospatial(fieldName, g, multiLineStringSchema));
+          break;
+        case MULTIPOLYGON:
+          Schema.Field multiPolygonsField = schema.getField(SapODataConstants.GEO_COLLECTION_MULTI_POLYGONS_FIELD_NAME);
+          Schema multiPolygonSchema = multiPolygonsField.getSchema().getComponentSchema();
+          multiPolygons.add(extractGeospatial(fieldName, g, multiPolygonSchema));
+          break;
+      }
+    });
+
+    return StructuredRecord.builder(schema)
+      .set(SapODataConstants.GEOSPATIAL_TYPE_FIELD_NAME, "GeometryCollection")
+      .set(SapODataConstants.GEOSPATIAL_DIMENSION_FIELD_NAME, collection.getDimension().name())
+      .set(SapODataConstants.GEO_COLLECTION_POINTS_FIELD_NAME, points)
+      .set(SapODataConstants.GEO_COLLECTION_LINE_STRINGS_FIELD_NAME, lineStrings)
+      .set(SapODataConstants.GEO_COLLECTION_POLYGONS_FIELD_NAME, polygons)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_POINTS_FIELD_NAME, multiPoints)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_LINE_STRINGS_FIELD_NAME, multiLineStrings)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_POLYGONS_FIELD_NAME, multiPolygons)
+      .build();
+  }
+
+  private StructuredRecord extractPoint(Point point, Schema schema) {
+    return StructuredRecord.builder(schema)
+      .set(SapODataConstants.GEOSPATIAL_DIMENSION_FIELD_NAME, point.getDimension().name())
+      .set(SapODataConstants.POINT_X_FIELD_NAME, point.getX())
+      .set(SapODataConstants.POINT_Y_FIELD_NAME, point.getY())
+      .set(SapODataConstants.POINT_Z_FIELD_NAME, point.getZ())
+      .build();
+  }
+
+  private StructuredRecord extractLineString(LineString lineString, Schema schema) {
```

Comment:
Is schema passed in because each schema has a different name? 

I think it makes more sense for each of the schemas to be constants. A LineString is a LineString, regardless of where it came from. It also would simplify the logic in this class, and make it less error prone.